### DISCUSSION TITLE
Reactions: Sort keys by time of first appearance

### DISCUSSION THREAD

**Discussion by jryans** [element-hq/element-meta#773] [Ideas]
...instead of by vote count.

**Comment by lampholder** [comment]
Are we certain we want this? Isn't vote count more informative?

**Comment by jryans** [comment]
I think @nadonomy's argument was that it's less surprising, allows you to build "messages" out of the emoji ordering (like 🛫🇬🇧🛬), and it means the emoji key won't move out from under you when you increment it.

Is that correct @nadonomy?

**Comment by nadonomy** [comment]
Correct. To expand a little; that user behaviour is already fairly well engrained from usage on other platforms, and it also allows early reactors (including the message sender themselves) to instigate what reactions they want to see more of— e.g. 🍾 on announcements. 

That, plus the current functionality of reactions dancing around your screen as you interact with them is unexpected and pretty frustrating.

There could be an argument for sorting by number of reactions when viewing older messages, but (a) that might break the 'storytelling' aspect of choosing what order to send reactions and (b) it makes it more complicated than it needs to be— figuring out the heuristics for when to switch e.g. after an amount of time, but only on timeline load so existing reactions don't unexpectedly change sort order etc so I'd prefer to look at that as a future refinement, if at all.

**Comment by lampholder** [comment]
SGTM

**Comment by jryans** [comment]
Added an [MSC comment](https://github.com/matrix-org/matrix-doc/pull/1849#pullrequestreview-244320323) to discuss whether the bundled reactions data will support this.

**Comment by daenney** [comment]
I find the current behaviour really jarring. If there's multiple emoji reactions and you click on one of them, it suddenly "teleports" away from under your cursor as the list reshuffles and a new emoji ends up under your cursor, leaving the impression you miss-clicked.

**Comment by anoadragon453** [comment]
(Reposting from https://github.com/vector-im/element-web/issues/22698)

I can confirm that this is a bit of fun communities very often take part in on Discord.

![image](https://user-images.githubusercontent.com/1342360/176683567-e3d0bf64-22f3-4498-9c37-9a12691c795b.png)
![image](https://user-images.githubusercontent.com/1342360/176684140-6c8d0281-3720-439a-ac6b-c6d9603514ad.png)
![image](https://user-images.githubusercontent.com/1342360/176684508-e51fcc6d-6f3f-4bfc-9f44-3a64899345b0.png)

Particularly in announcement channels, where users can only interact via emojis.

---

Interestingly, looking at the above screenshots you can see that even though reactions aren't sorted by count on Discord, you almost end up with that sorting occurring anyways. This is likely because (in left-to-right reading countries) people will start looking, and reacting, from the left-most reaction to the right.

So if one is looking to see the most upvoted answer, that may happen naturally anyways. Sacrificing original reaction order may not be worth it. This information could also possibly be exposed elsewhere, such as the proposed feature of a modal showing which users have reacted with which emoji.

A final argument. Re-ordering reactions is annoying when you're attempting to garner interest from others via reactions (or a bot is using them for user responses). A bot may post a message like the following:

> Choose the categories of events you'd like to be notified about!
>
> :microphone: Weekly tech talks.
> :video_game: Community game streams.
> :medal_military: Speed-running competitions.
> :movie_camera: Movie nights.
> :woman_cook: Group cooking nights.

Typically the bot would then react to the message with each emoji in the same order as the list above, so that it's easy to cross-reference. But with inconsistent reaction ordering, this can end up complicating the user experience.

![image](https://user-images.githubusercontent.com/1342360/176694072-367b13a6-e1e2-40cd-abc7-8d10c183cfcf.png)

**Comment by anoadragon453** [comment]
![image](https://user-images.githubusercontent.com/1342360/195886897-6fe26c3a-b70c-4882-a9c5-03eceb8053ea.png)

This behaviour has started to show up in the Synapse announcements room.

**Comment by babolivier** [comment]
> This behaviour has started to show up in the Synapse announcements room.

It's actually happened a few times now and has been consistently mangled by Element :frowning_face: 

![image](https://user-images.githubusercontent.com/5547783/195888755-d81b38b9-40d9-4773-8fe3-99cb8ed15f1e.png)
